### DISCUSSION TITLE
Toggle the TabSwitcher

### DISCUSSION THREAD

**Discussion by etiennelacoursiere** [zed-industries/zed#10933] [Feature Requests]
The TabSwitcher work differently than `cmd+p` ( search filles ). Instead of opening a picker and closing it when you select, you need to keep pressing tab and the selection happen when you release `ctrl`. 

Is there a way to have the tab switcher act like `cmd-p` ? Meaning that it will stay open until you select something? 

So to be clear, i would want to hit a keybind, open the tab switcher, navigate it with the arrows or using `ctrl-k` and `ctrl-j` that i have mapped to `menu::SelectPrev` and `menu::SelectNext` then select a tab with enter.

**Comment by alygin** [comment]
No, it's not possible at the moment.

But I wonder why would you want to do that? Your desired implementation always requires more keystrokes for the same result.

The file finder uses different approach because it has the search-by-text feature that requires keeping the window open. Unlike the tab switcher, it doesn't display all the options right away. Yet it also supports the quick selection mode &mdash; if you keep pressing `cmd-p`, it will switch to the selected file as soon as you release `cmd`.

BTW, you can use `ctrl-k` / `ctrl-j` and arrows to navigate items in the tab switcher, and confirm the selection with `enter`. Just hold `ctrl` while you're doing it.

**Reply by etiennelacoursiere** [reply]
Thank you for your response. Didn't know i could navigate with `ctrl-k` / `ctrl-j` that's great.

Why i wanted to do that. I'm just super use to neovim and telescope and i want to have a similar experience. Trying to reduce the friction of switching.  

Also, I like to be able to search textually in my open buffers in neovim. I find it faster to open the buffers finder, see my open buffers and start typing the name of the buffer instead of navigating to it. 

I just realized that `cmd+p` seems to give me my open tabs so thats exactly what i want.

**Comment by baldwindavid** [comment]
Worth noting that you can get this behavior by just running the command `tab_switcher::Toggle`, which you can attach a keybinding to.

I've been experimenting with using both the quick selection (`cmd-i`) and manual (`space i`). I've added `j k l` movement keys for the manual method; arrow keys also work due to defaults. Here are my relevant mappings...

```json
  {
    "bindings": {
      "cmd-i": "tab_switcher::Toggle"
    }
  },

  {
    "context": "Editor && VimControl && !VimWaiting && !menu",
    "bindings": {
      "space i": "tab_switcher::Toggle"
    }
  },

  {
    "context": "TabSwitcher",
    "bindings": {
      "k": "menu::SelectPrev",
      "j": "menu::SelectNext",
      "l": "menu::Confirm",
      "x": "tab_switcher::CloseSelectedItem",

      "cmd-shift-i": "menu::SelectPrev",
      "cmd-backspace": "tab_switcher::CloseSelectedItem"
    }
  },
```

**Reply by alygin** [reply]
You're right, it is possible to keep the TabSwitcher open in Vim mode, I didn't think of that.

When the TabSwitcher opens, it remembers the pressed modifier keys. When those modifier keys are released, the TabSwitcher confirms the current selection. So, if no modifiers are used to open the switcher, it'll stay open until an explicit confirmation or cancellation.

**Reply by LostKobrakai** [reply]
Hmm. It would be great to have more control over the held modifier key. 

The setup I'm using is `cmd+e` then then use `cmd+e` to move downwards in the list and `cmd-shift-e` to move upwards. However I also want to be able to start with `cmd-shift-e` to start at the end of the list, but now I can no longer navigate downwards (e.g. after some more `cmd-shift-e`) because to trigger `cmd-e` I need to lift shift, closing the tab switcher, even while holding cmd.

**Reply by nkabbara** [reply]
hey @baldwindavid, I"m curious how you found out about the TabSwitcher context. I was going with this list for valid contexts: https://zed.dev/docs/key-bindings, but it's not there.

**Reply by baldwindavid** [reply]
@nkabbara that was from poking around `zed: open default keymap` and `vim: open default keymap` from the command palette.

**Comment by CanRau** [comment]
Interestingly, arrow keys don't work for me when opening `tab_switcher::Toggle` via a keybinding, but with `tab_switcher::ToggleAll`, arrows work.
Would be nice to have arrows work with `tab_switcher::Toggle` as well, but I think I can get used to it without 🤓

**Reply by cameronraysmith** [reply]
I agree. How about just [overriding the shortcut](https://zed.dev/docs/key-bindings) `Toggle -> ToggleAll`

```jsonc
  {
    "context": "Workspace",
    "bindings": {
      "ctrl-tab": "tab_switcher::ToggleAll"
    }
  },
```

?

Seems to work well if you're also using settings.json with

```jsonc
  "tab_bar": {
    "show": false
  }
```

or similar.
